src/python/pants/backend/cc/dependency_inference/rules.py

```diff
@@ -65,10 +69,45 @@ class CCFilesMapping:
     mapping: FrozenDict[str, Address]
     ambiguous_files: FrozenDict[str, tuple[Address, ...]]
     mapping_not_stripped: FrozenDict[str, Address]
+    subdirectory_mapping: FrozenDict[str, Address]
+
+
+@rule_helper
+async def _subdirectory_mapping(
```

Comment:
This is the concept of SourceRoot + "include", which will become compiler include directories, but it lets us infer locations like `core/foo.h` located in: `core/include/core/foo.h` - which is a super standard way to locate public headers.

Those headers located in the `src` directory, next to the `.c` files are usually private to that library.

Reply:
Not particularly happy with the code here - feels like there should be a more elegant Pythonic solution

Reply:
cc @Eric-Arellano